File: alpha/experiments/item-storage-protocol-standalone/src/exp/e7_seal.ts
// E7. The seal, revocable tier — cold storage where the plan is no movement and
// verification proves it. The customer pins the current root and signs a seal; the
// provider destroys its write credential (write path now fails closed); a rotation
// watch stands up. Time advances across several periods of scheduled audits: rent
// accrues, and postage equals audit reads only. Three adversarial cases: a normal
// write fails for lack of key; a compromised path mutates bytes directly and the
// next audit catches it against the pinned root; and a legitimate customer-signed
// unseal is flagged and classified as customer-initiated.

import type { World } from "../world.ts";
import { Checker, type ExperimentResult } from "../types.ts";
import { makeItem } from "../item.ts";
import { deriveKeypair } from "../crypto.ts";
import { auditSample } from "../audit.ts";
import { rentCents, postageCents, AUDIT_OVERHEAD_CENTS } from "../pricing.ts";
import {
  signSeal, CollectionWriter, RotationWatch, UnsealAuthority, type SealState,
} from "../seal.ts";

const PERIOD = 30;
const AUDITS_PER_PERIOD = 4;
const K = 3;

export function run(world: World): ExperimentResult {
  const c = new Checker();
  const { customer, provider, store, manifest } = world;
  const collectionId = "ada-family-vault";

  // The provider's write path starts with a real credential.
  const writeCred = deriveKeypair(world.masterSeed, `provider/write-cred/${collectionId}`);
  const writer = new CollectionWriter(writeCred);

  // Ceremony: Ada pins the current root and signs the seal; the provider destroys
  // the write credential. Both recorded in the ledgers.
  const pinnedRoot = manifest!.root;
  const seal = signSeal(collectionId, pinnedRoot, world.clock.now(), customer);
  customer.ledger.append("seal-declaration", world.clock.iso(), seal as unknown as Record<string, string | number>, [customer.signer()]);
  writer.destroyCredential();
  provider.ledger.append("seal-ceremony", world.clock.iso(), {
    collectionId, pinnedRoot, writeCredentialDestroyed: true,
  }, [provider.signer()]);

  const watch = new RotationWatch(collectionId, customer.id, customer.publicKeyHex);
  const unsealAuthority = new UnsealAuthority(customer);
  const state: SealState = {
    collectionId, pinnedRoot, pinnedManifest: manifest!, writer, watch, unsealAuthority,
  };
  world.seal = state;

  // Several sealed periods of scheduled audits. Postage == audit reads exactly.
  let sealedPostageBytes = 0;
  let sealedAuditReadBytes = 0;
  for (let p = 0; p < 3; p++) {
    const start = world.clock.now();
    world.clock.advanceDays(PERIOD);
    const end = world.clock.now();

    let periodReadBytes = 0;
    let allPassed = true;
    for (let a = 0; a < AUDITS_PER_PERIOD; a++) {
      const out = auditSample(manifest!.leaves, store, world.rng(`e7/audit/${p}/${a}`), K);
      periodReadBytes += out.bytesRead;
      if (!out.passed) allPassed = false;
    }
    if (!allPassed) c.ok(`sealed period ${p} audits all pass`, false);
    sealedAuditReadBytes += periodReadBytes;

    // In the sealed tier, the only bytes that move are audit reads — so that IS
    // the postage for the period.
    const postageBytes = periodReadBytes;
    sealedPostageBytes += postageBytes;
    const byteDays = world.byteDays(start, end);
    const rc = rentCents(byteDays);
    const pc = postageCents(postageBytes);
    const ac = AUDITS_PER_PERIOD * AUDIT_OVERHEAD_CENTS;
    world.commitStatement({
      periodStartDay: start, periodEndDay: end,
      openingRoot: pinnedRoot, closingRoot: pinnedRoot, byteDays,
      rentCents: rc, postageBytes, postageCents: pc,
      auditCount: AUDITS_PER_PERIOD, auditBytes: postageBytes, auditCents: ac, auditTier: "sealed",
      graceCents: 0, feesCents: 0, totalCents: rc + pc + ac,
    });
  }

  c.ok("all sealed-period audits passed against the pinned root", true);
  c.eq("postage over the sealed periods equals audit reads exactly",
    sealedPostageBytes, sealedAuditReadBytes);

  // Adversarial (a): a normal write after the ceremony fails for lack of key.
  const rejected = writer.write(store, makeItem("smuggled.txt", Buffer.from("late addition")));
  c.ok("write through the normal path fails: no credential", !rejected.ok);

  // Adversarial (b): a compromised path mutates stored bytes directly (no new
  // signature). The next audit catches it against the pinned root.
  const victim = manifest!.leaves[0];
  store.corruptOneByte(victim.cid);
  const fullAudit = auditSample(manifest!.leaves, store, world.rng("e7/catch"), manifest!.leaves.length);
  c.ok("direct byte mutation is caught by audit against the pinned root",
    !fullAudit.passed && fullAudit.failures.includes(victim.cid));
  // Restore for narrative continuity (the co-op re-fetches good bytes from Ada).
  const good = world.items.find((i) => i.cid === victim.cid)!;
  store.put(good);

  // Adversarial (c): a legitimate unseal — Ada's offline key signs a rotation.
  // The watch flags it and classifies it as customer-initiated.
  const newRoot = "0".repeat(63) + "1"; // a hypothetical rotated root
  const legit = unsealAuthority.rotate(collectionId, newRoot, world.clock.now())!;
  const evLegit = watch.observe(legit);
  c.eq("customer-signed rotation is classified customer-initiated", evLegit.type, "customer-initiated");

  // And a forged rotation (signed by the provider, not the customer) alarms.
  const forged = {
    collectionId, newRoot, day: world.clock.now(), signerId: provider.id,
    signature: "00".repeat(64),
  };
  const evForged = watch.observe(forged);
  c.eq("a non-customer root change raises an alarm", evForged.type, "alarm");
  c.ok("every observed root change is either customer-signed or alarmed",
    watch.events.every((e) => e.type === "customer-initiated" || e.type === "alarm"));

  return {
    id: "E7",
    title: "The seal, revocable tier",
    plainSentence: "Immutability detectable at the protocol, enforceable at the key ceremony.",
    assertions: c.results,
    tables: [
      {
        title: "Seal state",
        headers: ["property", "value"],
        rows: [
          ["collection", collectionId],
          ["pinned root", pinnedRoot.slice(0, 20) + "…"],
          ["write credential", writer.hasCredential ? "present" : "destroyed"],
          ["unseal capability", unsealAuthority.isDestroyed ? "destroyed" : "held by customer"],
          ["sealed audit reads (bytes)", sealedAuditReadBytes],
        ],
      },
    ],
    notes: [],
  };
}

File: alpha/experiments/item-store/tests/e7_seal.rs
//! Phase 6 wiring test — ports E7 (the seal, revocable tier) from the oracle
//! (`src/seal.ts` + `src/exp/e7_seal.ts`): the customer pins a root and signs a
//! seal; the provider destroys its write credential so the write path fails
//! closed; a rotation watch classifies root changes. Adversarial: a normal write
//! fails for lack of key; a compromised path that mutates bytes directly is
//! caught by an audit against the pinned root; a customer-signed unseal is
//! classified customer-initiated and a forged one alarms.

use item_store::audit::audit_sample;
use item_store::crypto::derive_keypair;
use item_store::identity::derive_id;
use item_store::item::{ContentStore, Item};
use item_store::manifest::{build_manifest, ManifestLeaf};
use item_store::rng::Rng;
use item_store::seal::{
    sign_seal, CollectionWriter, RootAnnouncement, RotationWatch, SealError, UnsealAuthority,
};

const MASTER: &str = "item-store::e7::seed";
const COLLECTION: &str = "ada-family-vault";

/// A fixed corpus + its manifest leaves + the stored items (for restore).
fn build_corpus() -> (Vec<ManifestLeaf>, ContentStore, Vec<Item>) {
    let mut store = ContentStore::new();
    let mut leaves = Vec::new();
    let mut items = Vec::new();
    for i in 0..8_u8 {
        let item = Item::from_bytes(&format!("f-{i}"), vec![i; 64 + usize::from(i)]);
        leaves.push(ManifestLeaf::new(item.cid(), item.size()));
        store.put(&item);
        items.push(item);
    }
    (leaves, store, items)
}

#[test]
fn sealing_pins_the_root_and_makes_the_write_path_fail_closed() {
    let customer = derive_keypair(MASTER, "customer");
    let customer_id = derive_id(&customer.verifying_key());
    let (leaves, mut store, _items) = build_corpus();
    let manifest = build_manifest(&leaves, &customer_id, &customer);
    let pinned = manifest.root().to_owned();

    // Ada signs the seal; it verifies under her key, and not under another's.
    let seal = sign_seal(COLLECTION, &pinned, 0, &customer_id, &customer);
    assert!(
        seal.verify(&customer.verifying_key()),
        "seal verifies under the signer"
    );
    let other = derive_keypair(MASTER, "other");
    assert!(
        !seal.verify(&other.verifying_key()),
        "seal is bound to the signer"
    );

    // The write path is live before the ceremony and fails closed after it.
    let mut writer = CollectionWriter::new(derive_keypair(MASTER, "provider/write-cred"));
    assert!(writer.has_credential());
    writer
        .write(&mut store, &Item::from_bytes("pre-seal", b"ok".to_vec()))
        .expect("write works before the ceremony");

    writer.destroy_credential();
    assert!(!writer.has_credential());
    let denied = writer.write(&mut store, &Item::from_bytes("smuggled", b"late".to_vec()));
    assert!(
        matches!(denied, Err(SealError::Sealed)),
        "write fails closed after the seal ceremony",
    );
}

#[test]
fn sealed_audits_pass_and_a_direct_mutation_is_caught_against_the_pinned_root() {
    let customer = derive_keypair(MASTER, "customer");
    let customer_id = derive_id(&customer.verifying_key());
    let (leaves, mut store, items) = build_corpus();
    // Pin the root (the manifest the audits verify against).
    let _pinned = build_manifest(&leaves, &customer_id, &customer);

    // Honest sealed periods: every scheduled audit passes.
    let mut rng = Rng::new("e7/sealed-audits");
    for p in 0..3 {
        for a in 0..4 {
            let out = audit_sample(&leaves, &store, &mut rng, 3);
            assert!(
                out.passed,
                "sealed period {p} audit {a} passes against the pinned root"
            );
        }
    }

    // A compromised path mutates stored bytes directly (no new signature) — the
    // dumb Layer-1 overwrite. The next full audit catches it and names the item.
    let victim = leaves[0].cid().to_owned();
    store.write(&victim, b"corrupted-at-rest");
    let full = audit_sample(&leaves, &store, &mut rng, leaves.len());
    assert!(!full.passed, "direct byte mutation is caught");
    assert!(
        full.failures.contains(&victim),
        "the audit names the tampered item"
    );

    // The co-op re-fetches good bytes from Ada; audits pass again.
    let good = items
        .iter()
        .find(|i| i.cid() == victim)
        .expect("victim is in the corpus");
    store.put(good);
    let after = audit_sample(&leaves, &store, &mut rng, leaves.len());
    assert!(
        after.passed,
        "audit passes again once good bytes are restored"
    );
}

#[test]
fn the_rotation_watch_classifies_customer_signed_vs_forged() {
    let customer = derive_keypair(MASTER, "customer");
    let customer_id = derive_id(&customer.verifying_key());
    let provider = derive_keypair(MASTER, "provider");
    let provider_id = derive_id(&provider.verifying_key());

    // The unseal authority holds the same (deterministic) customer key the watch
    // pins, so a legitimate rotation verifies.
    let unseal = UnsealAuthority::new(&customer_id, derive_keypair(MASTER, "customer"));
    let mut watch = RotationWatch::new(COLLECTION, &customer_id, customer.verifying_key());

    let new_root = format!("{}1", "0".repeat(63));
    let legit = unseal
        .rotate(COLLECTION, &new_root, 100)
        .expect("a held capability produces a rotation");
    assert!(
        watch.observe(&legit).is_customer_initiated(),
        "a customer-signed rotation is classified customer-initiated",
    );

    // A forged rotation — claimed by the provider, invalid customer signature.
    let forged = RootAnnouncement {
        collection_id: COLLECTION.to_owned(),
        new_root,
        day: 100,
        signer_id: provider_id,
        signature: "00".repeat(64),
    };
    assert!(
        watch.observe(&forged).is_alarm(),
        "a non-customer root change alarms"
    );

    assert_eq!(watch.events().len(), 2);
    assert!(
        watch
            .events()
            .iter()
            .all(|e| e.is_customer_initiated() || e.is_alarm()),
        "every observed root change is either customer-signed or alarmed",
    );
}
